### ISSUE TITLE
Contact form endpoint

### ISSUE THREAD

**Issue by tisto** [plone/plone.restapi#457]
plone.restapi currently does not support the Plone contact form use case. We need an endpoint that allows anonymous users to contact the site owners (name, from, subject, message)

**Comment by tisto** [comment]
Draft of a very basic implementation:

```
POST http://localhost:8080/Plone/@contact-info
Accept: application/json
Content-Type: application/json

{
  'name': 'John Doe',
  'from': 'john@doe.com',
  'subject': 'Hello!',
  'message': 'Just want to say hi.'
} 
```

I am not sure about the URL '@contact-info' is what Plone uses for the default contact form. '@feedback', '@contact', '@email'? Opinions @sneridagh @robgietema @jaroel?

To reach a particular user, we could use a GET param "user". Since we do not have user content objects, we can not make the endpoint context sensitive. Not sure if that functionality should go into plone.restapi at all though.

**Comment by sneridagh** [comment]
Depends on how email-ish do we want it to be like or if we want to restrict the use case to the contact form.

If at the end we want a way to communicate between external or internal users (let's suppose we have an internal messaging system) I would go generic (without making it email-ish) and call it:

**@message**
```
{
  'name': 'John Doe',
  'from': 'john@doe.com',
  'to': 'jane@doe.com',
  'message': 'Just want to say hi.'
} 
```
Do we really need "subject" field? 

If we want to focus on the contact use case, then I would go for the:

**@contact**
```
{
  'name': 'John Doe',
  'from': 'john@doe.com',
  'message': 'Just want to say hi.'
} 
```
No "subject", and no recipient.

If we ship this, we have to make sure that we provide documentation to people with a way to prevent flooding via Nginx or so.

**Comment by davisagli** [comment]
An internal messaging system should work by specifying the user id, not the recipient's email. That way the recipient's email doesn't need to be provided to the frontend if it shouldn't be revealed, and it makes it a bit harder to use for sending spam.

**Comment by jaroel** [comment]
Internal messaging and contact forms aren't really the same. The first might send notification emails, the latter (mostly) is just an email to the site owner.

I'm more then happy to work on a proper messaging solution, which we'll probably have to extract from Quaive (or Castle if it has that).

As for the contact form, I'd think this should be sufficient:
```
POST http://localhost:8080/Plone/@contact-form
Accept: application/json
Content-Type: application/json

{
  <optional>'name': 'John Doe',
  'from': 'john@doe.com',
 <optional> 'subject': 'Hello!',
  'message': 'Just want to say hi.'
} 
```
Note the "contact-**form**"

**Comment by tisto** [comment]
We currently have two use cases:

1) Contact site owner
2) Contact users of the site ( without exposing their email addresses)

This is not about messaging between users. 

Both options can not set the "to" email addresses. We just send emails to pre-configured users. 

I am ok with just implementing 1) in plone.restapi and do 2) in our project if the two solutions differ too much. 

Von meinem iPhone gesendet

**Comment by jaroel** [comment]
We'll have to keep in mind that when we're using email-as-login that the userids most probably also is the email address.

**Comment by tisto** [comment]
@jaroel right, this makes the second use case harder to implement. We definitely do not want to expose the email address of the user. We also might have a third use case: send a content object to a user. This has been removed from the Plone UI some time ago so I am not sure if we still have to support that use case. I can imagine this only makes sense to allow to logged in users to prevent abuse.

**Comment by tisto** [comment]
@jaroel we might have to obfuscate the user ids anyways. The "@users" endpoint is currently protected by "manage portal" permissions. Though, if we ever want to make this less restrict, we might have to obfuscate our user ids. So what if we would allow the users endpoint to work with obfuscated user ids as well, e.g.:

GET localhost:8080/Plone/@users/7s88Df

(not sure if we should use UUIDs or our own schema here, needs some more research).

If we have that, we can easily implement the "contact-form" or "email-notification" endpoint on top of the user resource:

POST localhost:8080/Plone/@users/7s88Df/email-notification

What do you think?

**Comment by tisto** [comment]
I created a pull request with the initial documentation: https://github.com/plone/plone.restapi/pull/461

**Comment by tisto** [comment]
Ok, I just found that there is already a setting in the Plone security control panel to obfuscate the userid with a UUID:

> Use UUID user ids Use automatically generated UUIDs as user id for new users. When not turned on, the default is to use the same as the login name, or when using the email address as login name we generate a user id based on the fullname.

When enabled, newly registered users get a UUID as user id and this is reflected in the plone.restapi call. To get the user information you just do a GET request on:

`http://<SERVER>/api/@users/1ee6df1545444dd7bf487e5caf17ded0`

Therefore we can just implement it as proposed and leave it to the site administrator to enable this setting when using email logins (and add a note to the plone.restapi docs).

**Comment by tisto** [comment]
Implemented and merged in #461.
